### DISCUSSION TITLE
How to convert from array serialization to scalar value?

### DISCUSSION THREAD

**Discussion by reypm** [symfony/symfony#42752] [Q&A]
I have the following DTO:

```php
<?php

declare(strict_types=1);

namespace App\DTO;

use App\DTO\DTOInterface;
use App\DTO\ElectronicDocumentDTO;
use App\DTO\GroupValidation;
use JMS\Serializer\Annotation as Serializer;
use Symfony\Component\Validator\Constraints\DateTime;
use Symfony\Component\Validator\Constraints\NotBlank;

/**
 * @Serializer\ExclusionPolicy("NONE")
 */
class CreateCreditNoteDTO extends ElectronicDocumentDTO implements DTOInterface
{
    /**
     * @Serializer\Type("string")
     */
    #[NotBlank]
    #[DateTime(format: 'Y-m-d H:i:s')]
    public string $date;

    /**
     * @Serializer\Type("array<App\DTO\Base\DetailDTO>")
     */
    #[NotBlank]
    #[Valid]
    public array $details;
}
```
`$details` holds a bunch of other properties/fields that are also serialized. I am also benchmarking this code and have notice the following:
 - with `@Serializer\Type("array<App\DTO\Base\DetailDTO>` the Serializer is called a lot of times increasing the time the script takes to execute: ~200-400 ms
 - without it the times are really good and reasonable: less than 90 ms

Is there any way to avoid this kind of serialization here but still getting valid values inside the `$details` variable? I was thinking in use something like:
```php
    /**
     * @var array
     *
     * @Assert\Type("array")
     * @Assert\All({
     *     @Assert\Type("integer")
     * })
     */
```
but I don't know how to check each property needed because not all of them are integers, ie (these are the values inside of the `DetailDTO` class):
```php
    /**
     * @Serializer\Type("string")
     */
    #[NotBlank(groups: ['invoice'])]
    #[Length(max: 25, groups: ['invoice', 'referralGuide'])]
    public string $mainCode;

    /**
     * @Serializer\Type("string")
     */
    #[NotBlank(groups: ['invoice'])]
    #[Length(max: 25, groups: ['invoice', 'referralGuide'])]
    public string $auxCode;

    /**
     * @Serializer\Type("string")
     */
    #[NotBlank(groups: ['invoice', 'creditNote', 'referralGuide'])]
    #[Length(max: 300)]
    public string $description;
```

**Comment by Nek-** [comment]
Hello, you should probably take a look at custom normalizers. https://symfony.com/doc/current/serializer/custom_normalizer.html

**Reply by reypm** [reply]
I don't think that would help so much with my issue 😐 also notice we are using JMS Serializer and not Symfony Seliarizer which makes this more difficult

**Reply by Nek-** [reply]
Oh sorry. As this forum is about Symfony only I thought you were asking for the Symfony serializer. My bad.

**Comment by norkunas** [comment]
Maybe [Collection](https://symfony.com/doc/current/reference/constraints/Collection.html) constraint would fit for you in this case?
